src/scipp/curve_fit.py

```diff
@@ -144,43 +127,38 @@ def _make_defaults(f, coords, params):
         **(spec.kwonlydefaults or {}),
     )
     return {
-        **{a: 1.0 for a in all_args - set(coords)},
+        **{a: scalar(1.0) for a in all_args - set(coords)},
         **default_arguments,
-        **(params or {}),
+        **{
+            k: v if isinstance(v, Variable) else scalar(v)
+            for k, v in (p0 or {}).items()
+        },
     }
 
 
-def _get_specific_bounds(bounds, name, unit) -> tuple[float, float]:
+def _get_specific_bounds(bounds, name, unit):
     if name not in bounds:
-        return -np.inf, np.inf
+        return -scalar(np.inf), scalar(np.inf)
     b = bounds[name]
     if len(b) != 2:
         raise ValueError(
             "Parameter bounds must be given as a tuple of length 2. "
             f"Got a collection of length {len(b)} as bounds for '{name}'."
         )
-    if isinstance(b[0], Variable):
-        return (
-            b[0].to(unit=unit, dtype=float).value,
-            b[1].to(unit=unit, dtype=float).value,
-        )
-    return b
+    le, ri = (v if isinstance(v, Variable) else scalar(v) for v in b)
```

Comment:
This is odd. Previously we required that either both bounds are variables or both are numbers. Now they can be mixed. Normally, we don't allow this. See, e.g., `linspace` and `arange` which can take either variables or numbers, but never a mixture.

I think `curve_fit` should not deviate from this.

What's more, `scipp.scipy.optimize.curve_fit` does not allow this. It will only lead to confusion when the two functions deviate from each other.

Reply:
I agree it's better to have them both be scalars or both be variables.

There's also a bug here, if `inf` is passed as the bound, you will get a unit error unless the associated parameter is dimensionless.

Reply:
I wouldn't call this a bug. This is consistent with how we treat raw numbers elsewhere. They are treated as dimensionless.